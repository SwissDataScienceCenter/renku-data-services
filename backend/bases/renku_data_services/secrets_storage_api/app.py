"""Secrets storage app."""

from sanic import Sanic

from renku_data_services.base_api.error_handler import CustomErrorHandler
from renku_data_services.base_api.misc import MiscBP
from renku_data_services.secrets import apispec
from renku_data_services.secrets.blueprints import K8sSecretsBP
from renku_data_services.secrets_storage_api.dependencies import DependencyManager


def register_all_handlers(app: Sanic, dm: DependencyManager) -> Sanic:
    """Register all handlers on the application."""
    url_prefix = "/api/secrets"
    secrets_storage = K8sSecretsBP(
        name="secrets_storage_api",
        url_prefix=url_prefix,
        user_secrets_repo=dm.user_secrets_repo,
        authenticator=dm.authenticator,
        secret_service_private_key=dm.config.secrets.private_key,
        previous_secret_service_private_key=dm.config.secrets.previous_private_key,
        core_client=dm.core_client,
    )
    misc = MiscBP(name="misc", url_prefix=url_prefix, apispec=dm.config.spec, version=dm.config.version)
    app.blueprint([secrets_storage.blueprint(), misc.blueprint()])

    app.error_handler = CustomErrorHandler(apispec)
    app.config.OAS = False
    app.config.OAS_UI_REDOC = False
    app.config.OAS_UI_SWAGGER = False
    app.config.OAS_AUTODOC = False
    return app
