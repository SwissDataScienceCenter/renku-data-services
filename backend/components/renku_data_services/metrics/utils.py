"""Utility functions."""

import hashlib

from renku_data_services.base_models.core import APIUser


def anonymize_user_id(user: APIUser) -> str:
    """Anonymize a user's id."""
    return (
        hashlib.md5(user.id.encode("utf-8"), usedforsecurity=False).hexdigest() if user.id is not None else "anonymous"
    )
