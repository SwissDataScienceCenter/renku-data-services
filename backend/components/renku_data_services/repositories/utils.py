"""Utilities for connected services."""

import httpx


async def probe_repository(repository_url: str) -> bool:
    """Probe a repository to check if it is publicly available."""
    async with httpx.AsyncClient(timeout=5) as client:
        url = f"{repository_url}/info/refs?service=git-upload-pack"
        try:
            res = await client.get(url=url, follow_redirects=True)
            if res.status_code != 200:
                return False
            return bool(res.headers.get("Content-Type") == "application/x-git-upload-pack-advertisement")
        except httpx.HTTPError:
            return False
