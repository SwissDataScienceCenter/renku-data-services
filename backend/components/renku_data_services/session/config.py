"""Configuration for session module."""

import os
from dataclasses import dataclass
from datetime import timedelta

from pydantic import ValidationError as PydanticValidationError

from renku_data_services.app_config import logging
from renku_data_services.session import crs as session_crs

logger = logging.getLogger(__name__)


@dataclass
class BuildsConfig:
    """Configuration for container image builds."""

    enabled: bool = False
    build_output_image_prefix: str | None = None
    vscodium_python_run_image: str | None = None
    build_strategy_name: str | None = None
    push_secret_name: str | None = None
    buildrun_retention_after_failed: timedelta | None = None
    buildrun_retention_after_succeeded: timedelta | None = None
    buildrun_build_timeout: timedelta | None = None
    node_selector: dict[str, str] | None = None
    tolerations: list[session_crs.Toleration] | None = None

    @classmethod
    def from_env(cls) -> "BuildsConfig":
        """Create a config from environment variables."""
        enabled = os.environ.get("IMAGE_BUILDERS_ENABLED", "false").lower() == "true"
        build_output_image_prefix = os.environ.get("BUILD_OUTPUT_IMAGE_PREFIX")
        vscodium_python_run_image = os.environ.get("BUILD_VSCODIUM_PYTHON_RUN_IMAGE")
        build_strategy_name = os.environ.get("BUILD_STRATEGY_NAME")
        push_secret_name = os.environ.get("BUILD_PUSH_SECRET_NAME")
        buildrun_retention_after_failed_seconds = int(os.environ.get("BUILD_RUN_RETENTION_AFTER_FAILED_SECONDS") or "0")
        buildrun_retention_after_failed = (
            timedelta(seconds=buildrun_retention_after_failed_seconds)
            if buildrun_retention_after_failed_seconds > 0
            else None
        )
        buildrun_retention_after_succeeded_seconds = int(
            os.environ.get("BUILD_RUN_RETENTION_AFTER_SUCCEEDED_SECONDS") or "0"
        )
        buildrun_retention_after_succeeded = (
            timedelta(seconds=buildrun_retention_after_succeeded_seconds)
            if buildrun_retention_after_succeeded_seconds > 0
            else None
        )
        buildrun_build_timeout_seconds = int(os.environ.get("BUILD_RUN_BUILD_TIMEOUT") or "0")
        buildrun_build_timeout = (
            timedelta(seconds=buildrun_build_timeout_seconds) if buildrun_build_timeout_seconds > 0 else None
        )

        if os.environ.get("DUMMY_STORES", "false").lower() == "true":
            enabled = True  # Enable image builds when running tests

        node_selector: dict[str, str] | None = None
        node_selector_str = os.environ.get("BUILD_NODE_SELECTOR")
        if node_selector_str:
            try:
                node_selector = session_crs.NodeSelector.model_validate_json(node_selector_str).root
            except PydanticValidationError:
                logger.error("Could not validate BUILD_NODE_SELECTOR. Will not use node selector for image builds.")

        tolerations: list[session_crs.Toleration] | None = None
        tolerations_str = os.environ.get("BUILD_NODE_TOLERATIONS")
        if tolerations_str:
            try:
                tolerations = session_crs.Tolerations.model_validate_json(tolerations_str).root
            except PydanticValidationError:
                logger.error("Could not validate BUILD_NODE_TOLERATIONS. Will not use tolerations for image builds.")

        return cls(
            enabled=enabled or False,
            build_output_image_prefix=build_output_image_prefix or None,
            vscodium_python_run_image=vscodium_python_run_image or None,
            build_strategy_name=build_strategy_name or None,
            push_secret_name=push_secret_name or None,
            buildrun_retention_after_failed=buildrun_retention_after_failed,
            buildrun_retention_after_succeeded=buildrun_retention_after_succeeded,
            buildrun_build_timeout=buildrun_build_timeout,
            node_selector=node_selector,
            tolerations=tolerations,
        )
