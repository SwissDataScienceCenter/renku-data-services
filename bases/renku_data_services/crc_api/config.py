"""Configurations."""
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, Optional

import httpx
import renku_data_services.resource_pool_models as models
from jwt import PyJWKClient
from renku_data_services.resource_pool_adapters import ResourcePoolRepository, UserRepository
from renku_data_services.k8s.clients import DummyCoreClient, DummySchedulingClient, K8sCoreClient, K8sSchedulingClient
from renku_data_services.k8s.quota import QuotaRepository
from renku_data_services import errors
from renku_data_services.users.credentials import KeycloakAuthenticator
from renku_data_services.users.dummy import DummyAuthenticator, DummyUserStore
from renku_data_services.users.keycloak import KcUserStore
from tenacity import retry, stop_after_attempt, stop_after_delay, wait_fixed
from yaml import safe_load

from renku_data_services.crc_api.server_options import (
    ServerOptions,
    ServerOptionsDefaults,
    generate_default_resource_pool,
)

import renku_data_services.crc_schemas


@retry(stop=(stop_after_attempt(20) | stop_after_delay(300)), wait=wait_fixed(2), reraise=True)
def _oidc_discovery(url: str, realm: str) -> Dict[str, Any]:
    url = f"{url}/realms/{realm}/.well-known/openid-configuration"
    res = httpx.get(url)
    if res.status_code == 200:
        return res.json()
    raise errors.ConfigurationError(message=f"Cannot successfully do OIDC discovery with url {url}.")


default_resource_pool = models.ResourcePool(
    name="default",
    classes=[
        models.ResourceClass(
            name="small",
            cpu=0.5,
            memory=1,
            max_storage=20,
            gpu=0,
            default=True,
        ),
        models.ResourceClass(
            name="large",
            cpu=1.0,
            memory=2,
            max_storage=20,
            gpu=0,
            default=False,
        ),
    ],
    quota=None,
    public=True,
    default=True,
)


@dataclass
class Config:
    """Configuration for the CRC service."""

    user_repo: UserRepository
    rp_repo: ResourcePoolRepository
    user_store: models.UserStore
    authenticator: models.Authenticator
    quota_repo: QuotaRepository
    spec: Dict[str, Any] = field(init=False, default_factory=dict)
    version: str = "0.0.1"
    app_name: str = "renku_crc"
    default_resource_pool_file: Optional[str] = None
    default_resource_pool: models.ResourcePool = default_resource_pool
    server_options_file: Optional[str] = None
    server_defaults_file: Optional[str] = None

    def __post_init__(self):
        spec_file = Path(renku_data_services.crc_schemas.__file__).resolve().parent / "api.spec.yaml"
        with open(spec_file, "r") as f:
            self.spec = safe_load(f)
        if self.default_resource_pool_file is not None:
            with open(self.default_resource_pool_file, "r") as f:
                self.default_resource_pool = models.ResourcePool.from_dict(safe_load(f))
        if self.server_defaults_file is not None and self.server_options_file is not None:
            with open(self.server_options_file, "r") as f:
                options = ServerOptions.parse_obj(safe_load(f))
            with open(self.server_defaults_file, "r") as f:
                defaults = ServerOptionsDefaults.parse_obj(safe_load(f))
            self.default_resource_pool = generate_default_resource_pool(options, defaults)

    @classmethod
    def from_env(cls):
        """Create a config from environment variables."""

        prefix = ""
        user_store: models.UserStore
        authenticator: models.Authenticator
        version = os.environ.get(f"{prefix}VERSION", "0.0.1")
        keycloak_url = None
        keycloak_realm = "Renku"
        server_options_file = os.environ.get("SERVER_OPTIONS")
        server_defaults_file = os.environ.get("SERVER_DEFAULTS")
        k8s_namespace = os.environ.get("K8S_NAMESPACE", "default")

        if os.environ.get(f"{prefix}DUMMY_STORES", "false").lower() == "true":
            async_sqlalchemy_url = os.environ.get(
                f"{prefix}ASYNC_SQLALCHEMY_URL", "sqlite+aiosqlite:///data_services.db"
            )
            sync_sqlalchemy_url = os.environ.get(f"{prefix}SYNC_SQLALCHEMY_URL", "sqlite:///data_services.db")
            authenticator = DummyAuthenticator(admin=True)
            user_always_exists = os.environ.get("DUMMY_USERSTORE_USER_ALWAYS_EXISTS", "true").lower() == "true"
            user_store = DummyUserStore(user_always_exists=user_always_exists)
            quota_repo = QuotaRepository(DummyCoreClient({}), DummySchedulingClient({}), namespace=k8s_namespace)
        else:
            pg_host = os.environ.get("DB_HOST", "localhost")
            pg_user = os.environ.get("DB_USER", "renku")
            pg_port = os.environ.get("DB_PORT", "5432")
            db_name = os.environ.get("DB_NAME", "renku")
            pg_password = os.environ.get("DB_PASSWORD")
            if pg_password is None:
                raise errors.ConfigurationError(
                    message="Please provide a database password in the 'DB_PASSWORD' environment variable."
                )
            async_sqlalchemy_url = f"postgresql+asyncpg://{pg_user}:{pg_password}@{pg_host}:{pg_port}/{db_name}"
            sync_sqlalchemy_url = f"postgresql+psycopg2://{pg_user}:{pg_password}@{pg_host}:{pg_port}/{db_name}"
            keycloak_url = os.environ.get(f"{prefix}KEYCLOAK_URL")
            if keycloak_url is None:
                raise errors.ConfigurationError(message="The Keycloak URL has to be specified.")
            keycloak_url = keycloak_url.rstrip("/")
            keycloak_realm = os.environ.get(f"{prefix}KEYCLOAK_REALM", "Renku")
            oidc_disc_data = _oidc_discovery(keycloak_url, keycloak_realm)
            jwks_url = oidc_disc_data.get("jwks_uri")
            if jwks_url is None:
                raise errors.ConfigurationError(
                    message="The JWKS url for Keycloak cannot be found from the OIDC discovery endpoint."
                )
            algorithms = os.environ.get(f"{prefix}KEYCLOAK_TOKEN_SIGNATURE_ALGS")
            if algorithms is None:
                raise errors.ConfigurationError(message="At least one token signature algorithm is requried.")
            algorithms_lst = [i.strip() for i in algorithms.split(",")]
            jwks = PyJWKClient(jwks_url)
            authenticator = KeycloakAuthenticator(jwks=jwks, algorithms=algorithms_lst)
            user_store = KcUserStore(keycloak_url=keycloak_url, realm=keycloak_realm)
            quota_repo = QuotaRepository(K8sCoreClient(), K8sSchedulingClient(), namespace=k8s_namespace)

        user_repo = UserRepository(sync_sqlalchemy_url=sync_sqlalchemy_url, async_sqlalchemy_url=async_sqlalchemy_url)
        rp_repo = ResourcePoolRepository(
            sync_sqlalchemy_url=sync_sqlalchemy_url, async_sqlalchemy_url=async_sqlalchemy_url
        )
        return cls(
            user_repo=user_repo,
            rp_repo=rp_repo,
            version=version,
            authenticator=authenticator,
            user_store=user_store,
            quota_repo=quota_repo,
            server_defaults_file=server_defaults_file,
            server_options_file=server_options_file,
        )
