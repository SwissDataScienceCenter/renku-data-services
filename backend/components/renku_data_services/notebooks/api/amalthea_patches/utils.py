"""Various utilities for patching sessions."""

from typing import Any, cast

from kubernetes import client

from renku_data_services.notebooks.config import NotebooksConfig


def get_certificates_volume_mounts(
    config: NotebooksConfig,
    etc_certs: bool = True,
    custom_certs: bool = True,
    read_only_etc_certs: bool = False,
) -> list[dict[str, Any]]:
    """The list of volume mounts for custom certificates."""
    volume_mounts = []
    etc_ssl_certs = client.V1VolumeMount(
        name="etc-ssl-certs",
        mount_path="/etc/ssl/certs/",
        read_only=read_only_etc_certs,
    )
    custom_ca_certs = client.V1VolumeMount(
        name="custom-ca-certs",
        mount_path=config.sessions.ca_certs.path,
        read_only=True,
    )
    if etc_certs:
        volume_mounts.append(etc_ssl_certs)
    if custom_certs:
        volume_mounts.append(custom_ca_certs)
    return cast(list[dict[str, Any]], client.ApiClient().sanitize_for_serialization(volume_mounts))
