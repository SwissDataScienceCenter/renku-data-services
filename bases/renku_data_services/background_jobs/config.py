"""Configurations for background jobs."""

import os
from collections.abc import Callable
from dataclasses import dataclass

from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker

from renku_data_services.authz.authz import Authz
from renku_data_services.authz.config import AuthzConfig
from renku_data_services.data_connectors.migration_utils import DataConnectorMigrationTool
from renku_data_services.errors import errors
from renku_data_services.message_queue.config import RedisConfig
from renku_data_services.message_queue.db import EventRepository
from renku_data_services.message_queue.redis_queue import RedisQueue
from renku_data_services.namespace.db import GroupRepository
from renku_data_services.project.db import ProjectRepository
from renku_data_services.users.db import UsersSync
from renku_data_services.users.kc_api import IKeycloakAPI, KeycloakAPI


@dataclass
class SyncConfig:
    """Main configuration."""

    syncer: UsersSync
    kc_api: IKeycloakAPI
    authz_config: AuthzConfig
    group_repo: GroupRepository
    event_repo: EventRepository
    project_repo: ProjectRepository

    # NEW
    data_connector_migration_tool: DataConnectorMigrationTool

    session_maker: Callable[..., AsyncSession]

    @classmethod
    def from_env(cls, prefix: str = "") -> "SyncConfig":
        """Generate a configuration from environment variables."""
        pg_host = os.environ.get(f"{prefix}DB_HOST", "localhost")
        pg_user = os.environ.get(f"{prefix}DB_USER", "renku")
        pg_port = os.environ.get(f"{prefix}DB_PORT", "5432")
        db_name = os.environ.get(f"{prefix}DB_NAME", "renku")
        pg_password = os.environ.get(f"{prefix}DB_PASSWORD")
        if pg_password is None:
            raise errors.ConfigurationError(
                message="Please provide a database password in the 'DB_PASSWORD' environment variable."
            )
        async_sqlalchemy_url = f"postgresql+asyncpg://{pg_user}:{pg_password}@{pg_host}:{pg_port}/{db_name}"
        # NOTE: the pool here is not used to serve HTTP requests, it is only used in background jobs.
        # Therefore, we want to consume very few connections and we can wait for an available connection
        # much longer than the default 30 seconds. In our tests syncing 15 users times out with the default.
        engine = create_async_engine(async_sqlalchemy_url, pool_size=4, max_overflow=0, pool_timeout=600)
        session_maker: Callable[..., AsyncSession] = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)  # type: ignore[call-overload]
        redis = RedisConfig.from_env(prefix)
        message_queue = RedisQueue(redis)

        authz_config = AuthzConfig.from_env()
        event_repo = EventRepository(session_maker=session_maker, message_queue=message_queue)
        group_repo = GroupRepository(
            session_maker,
            event_repo=event_repo,
            group_authz=Authz(authz_config),
            message_queue=message_queue,
        )
        project_repo = ProjectRepository(
            session_maker=session_maker,
            message_queue=message_queue,
            event_repo=event_repo,
            group_repo=group_repo,
            authz=Authz(authz_config),
        )

        # NEW
        data_connector_migration_tool = DataConnectorMigrationTool(
            session_maker=session_maker,
            authz=Authz(authz_config),
        )

        syncer = UsersSync(
            session_maker,
            message_queue=message_queue,
            event_repo=event_repo,
            group_repo=group_repo,
            authz=Authz(authz_config),
        )
        keycloak_url = os.environ[f"{prefix}KEYCLOAK_URL"]
        client_id = os.environ[f"{prefix}KEYCLOAK_CLIENT_ID"]
        client_secret = os.environ[f"{prefix}KEYCLOAK_CLIENT_SECRET"]
        realm = os.environ.get(f"{prefix}KEYCLOAK_REALM", "Renku")
        kc_api = KeycloakAPI(keycloak_url=keycloak_url, client_id=client_id, client_secret=client_secret, realm=realm)
        return cls(
            syncer,
            kc_api,
            authz_config,
            group_repo,
            event_repo,
            project_repo,
            data_connector_migration_tool,
            session_maker,
        )
