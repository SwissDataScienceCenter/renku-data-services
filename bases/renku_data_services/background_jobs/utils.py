"""Different utility functions for background jobs."""

import logging

from authzed.api.v1.permission_service_pb2 import ReadRelationshipsRequest, RelationshipFilter

from renku_data_services.authz.authz import Authz, ResourceType, _Relation
from renku_data_services.background_jobs.config import SyncConfig
from renku_data_services.message_queue.avro_models.io.renku.events import v2
from renku_data_services.message_queue.converters import EventConverter


async def sync_user_namespaces(config: SyncConfig):
    """Lists all user namespaces in the database and adds them to Authzed and the event queue."""
    authz = Authz(config.authz_config)
    user_namespaces = await config.group_repo._get_user_namespaces()
    logging.info(f"Start syncing {len(user_namespaces)} to the authorization DB and message queue")
    for user_namespace in user_namespaces:
        events = EventConverter.to_events(user_namespace, v2.UserAdded)
        authz_change = authz._add_user_namespace(user_namespace.namespace)
        session = config.session_maker()
        tx = session.begin()
        await tx.start()
        try:
            await authz.client.WriteRelationships(authz_change.apply)
            for event in events:
                await config.event_repo.store_event(session, event)
        except Exception as err:
            logging.error(f"Failed to sync user namespace {user_namespace} because {err}")
            await tx.rollback()
        else:
            await tx.commit()
        finally:
            await session.close()
    logging.info(f"Wrote authorization changes and events for {len(user_namespaces)} user namespaces")


async def bootstrap_user_namespaces(config: SyncConfig):
    """Sycnhornize user namespaces to the authorization database only if none are already present."""
    authz = Authz(config.authz_config)
    rels = aiter(
        authz.client.ReadRelationships(
            ReadRelationshipsRequest(
                relationship_filter=RelationshipFilter(
                    resource_type=ResourceType.user_namespace.value, optional_relation=_Relation.owner.value
                )
            )
        )
    )
    num_rels = 0
    for _ in range(5):
        if await anext(rels, None) is not None:
            num_rels += 1
    if num_rels >= 5:
        logging.info(
            "Found at least 5 user namespace in the authorization database, "
            "will not sync user namespaces to authorization."
        )
        return
    await sync_user_namespaces(config)
