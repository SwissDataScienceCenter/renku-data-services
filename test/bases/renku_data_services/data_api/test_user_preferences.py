from test.bases.renku_data_services.data_api.utils import create_user_preferences
from typing import Any, Dict

import pytest
from sanic import Sanic
from sanic_testing.testing import SanicASGITestClient

from renku_data_services.data_api.app import register_all_handlers
from renku_data_services.data_api.config import Config

_valid_add_pinned_project: Dict[str, Any] = {"project_slug": "user.1/first-project"}


@pytest.fixture
def valid_add_pinned_project_payload() -> Dict[str, Any]:
    return _valid_add_pinned_project


@pytest.fixture
def test_client(app_config: Config) -> SanicASGITestClient:
    app = Sanic(app_config.app_name)
    app = register_all_handlers(app, app_config)
    return SanicASGITestClient(app)


@pytest.mark.asyncio
async def test_get_user_preferences(test_client: SanicASGITestClient, valid_add_pinned_project_payload: Dict[str, Any]):
    _, res = await create_user_preferences(test_client, valid_add_pinned_project_payload)
    assert res.status_code == 200

    _, res = await test_client.get(
        "/api/data/user_preferences",
        headers={"Authorization": "bearer test"},
    )

    assert res.status_code == 200
    assert res.json is not None
    assert res.json.get("user_id") == "some-id"
    assert res.json.get("pinned_projects") is not None
    assert len(res.json["pinned_projects"].get("project_slugs")) == 1
    project_slugs = res.json["pinned_projects"]["project_slugs"]
    assert project_slugs[0] == "user.1/first-project"
