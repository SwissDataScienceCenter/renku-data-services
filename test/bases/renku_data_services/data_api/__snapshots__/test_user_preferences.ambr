# serializer version: 1
# name: test_get_user_preferences
  dict({
    'pinned_projects': dict({
      'project_slugs': list([
        'user.1/first-project',
      ]),
    }),
  })
# ---
